# **Implementasi K-NN pada dataset iris (Tugas 3)**

## Konsep k-NN
k-NN (k-Nearest-Neighbors) merupakan sebuah algoritma yang digunakan untuk menentukan sebuah label dari data labeh yang sudah tersedia.
k-NN ini memeliki 5 algoritma yang digunakan
1.   Menentukan nilai K
2.   Menghitung jarak antara data uji dengan data latih (menggunakan Minkowski Distance)
3. Mengurutkan jarak dari terkecil hingga terbesar (Ascending)
4. Mengambil data sebanyak K terdekat
5. Memilih kelas

Menghitung jarak dengan cara data latih - data uji pada setiap fitur/label yang digunakan. kemudian jumlahkan.

Jika jumlah hasil label tersebut sama, maka bisa juga dipilih salah satu.




In [1]:
import pandas as pd
import numpy as np
pd.options.mode.chained_assignment = None 

## Mendapatkan dataset iris

In [2]:
dataset = pd.read_csv("https://gist.githubusercontent.com/netj/8836201/raw/6f9306ad21398ea43cba4f7d537619d0e07d5ae3/iris.csv")
dataset

sepal.length  sepal.width  petal.length  petal.width    variety
0             5.1          3.5           1.4          0.2     Setosa
1             4.9          3.0           1.4          0.2     Setosa
2             4.7          3.2           1.3          0.2     Setosa
3             4.6          3.1           1.5          0.2     Setosa
4             5.0          3.6           1.4          0.2     Setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  Virginica
146           6.3          2.5           5.0          1.9  Virginica
147           6.5          3.0           5.2          2.0  Virginica
148           6.2          3.4           5.4          2.3  Virginica
149           5.9          3.0           5.1          1.8  Virginica

[150 rows x 5 columns]

## Membuat dataframe
Label yang diguanakan adalah variety

In [3]:
X = dataset.drop(columns=['variety'])

In [4]:
y = dataset['variety'].values
y[0:29]

array(['Setosa', 'Setosa', 'Setosa', 'Setosa', 'Setosa', 'Setosa',
       'Setosa', 'Setosa', 'Setosa', 'Setosa', 'Setosa', 'Setosa',
       'Setosa', 'Setosa', 'Setosa', 'Setosa', 'Setosa', 'Setosa',
       'Setosa', 'Setosa', 'Setosa', 'Setosa', 'Setosa', 'Setosa',
       'Setosa', 'Setosa', 'Setosa', 'Setosa', 'Setosa'], dtype=object)

## Split dataset data uji dengan data latih
Ukuran data yang diguanakan adalah 20% dari data

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1, stratify=y)

## Membangun model
K yang digunakan sebanyak 15

In [6]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors = 15)
knn.fit(X_train,y_train)

KNeighborsClassifier(n_neighbors=15)

## Test model
Label yang diambil sebanyak 29 data yang sudah tersedia

In [7]:
knn.predict(X_test)[0:29]

array(['Virginica', 'Setosa', 'Versicolor', 'Setosa', 'Setosa', 'Setosa',
       'Virginica', 'Virginica', 'Virginica', 'Versicolor', 'Setosa',
       'Versicolor', 'Virginica', 'Versicolor', 'Virginica', 'Setosa',
       'Virginica', 'Versicolor', 'Versicolor', 'Virginica', 'Versicolor',
       'Versicolor', 'Setosa', 'Setosa', 'Virginica', 'Versicolor',
       'Setosa', 'Setosa', 'Versicolor'], dtype=object)

In [8]:
knn.score(X_test, y_test)

0.9666666666666667

In [10]:
from sklearn.model_selection import cross_val_score
import numpy as np

knn_cv = KNeighborsClassifier(n_neighbors=3)

cv_scores = cross_val_score(knn_cv, X, y, cv=5)

print(cv_scores)
print('cv_scores mean:{}'.format(np.mean(cv_scores)))

[0.96666667 0.96666667 0.93333333 0.96666667 1.        ]
cv_scores mean:0.9666666666666668


## Hasil

Hasil yang didapatkan adalah jika menggunakan data sebanyak 20%, labelnya sebanyak 30 dan nilai K yaitu , 3, 7, 10, 15 maka hasil scorenya tetap sama yaitu 0.9666666666666667